File: langs/python/src/vmx/capabilities/lifecycle_capabilities.py
"""Lifecycle capability contracts. See spec/14-capabilities.md §Lifecycle.

These three are baseline: every core VM trivially satisfies them.
"""

from __future__ import annotations

from abc import ABC, abstractmethod


class IConstructable(ABC):
    @abstractmethod
    def can_construct(self) -> bool: ...

    @abstractmethod
    def construct(self) -> None: ...


class IDestructable(ABC):
    @abstractmethod
    def can_destruct(self) -> bool: ...

    @abstractmethod
    def destruct(self) -> None: ...


class IReconstructable(ABC):
    @abstractmethod
    def can_reconstruct(self) -> bool: ...

    @abstractmethod
    def reconstruct(self) -> None: ...

File: langs/csharp/src/VMx/Capabilities/LifecycleCapabilities.cs
namespace VMx.Capabilities;

// Capability interfaces from spec/14-capabilities.md §Lifecycle.
// These three are baseline: every VM in the core library trivially satisfies
// them. See spec rule 2 in chapter 14 — the only capabilities the base VM
// types declare. Async variants follow ADR-0008 (C#-specific affordance).

/// <summary>Capability: the implementer has a construct lifecycle operation.</summary>
public interface IConstructable
{
    /// <summary>Returns true when <see cref="Construct"/> is valid to call.</summary>
    bool CanConstruct();

    /// <summary>Performs the construct operation.</summary>
    void Construct();

    /// <summary>
    /// Performs the construct operation asynchronously. Hook and deferred-child
    /// failures fault the task after transactional rollback.
    /// </summary>
    Task ConstructAsync();
}

/// <summary>Capability: the implementer has a destruct lifecycle operation.</summary>
public interface IDestructable
{
    /// <summary>Returns true when <see cref="Destruct"/> is valid to call.</summary>
    bool CanDestruct();

    /// <summary>Performs the destruct operation.</summary>
    void Destruct();

    /// <summary>
    /// Performs the destruct operation asynchronously. Hook and deferred-child
    /// failures fault the task after transactional rollback.
    /// </summary>
    Task DestructAsync();
}

/// <summary>Capability: the implementer has a reconstruct lifecycle operation.</summary>
public interface IReconstructable
{
    /// <summary>Returns true when <see cref="Reconstruct"/> is valid to call.</summary>
    bool CanReconstruct();

    /// <summary>Performs the reconstruct operation.</summary>
    void Reconstruct();

    /// <summary>
    /// Performs the reconstruct operation asynchronously. Hook and deferred-child
    /// failures fault the task after transactional rollback.
    /// </summary>
    Task ReconstructAsync();
}
